### ISSUE TITLE
Duplicate key Signature gradle-code-quality-tools-plugin:jar.asc:asc: with Gradle 5.1 & Java Gradle Plugin

### ISSUE THREAD

**Issue by vanniktech** [gradle/gradle#8213]
For publishing Jvm artifacts there has been the canonical script from Chris Banes - https://github.com/chrisbanes/gradle-mvn-push
I've wrapped that into a Gradle Plugin https://github.com/vanniktech/gradle-maven-publish-plugin

For both scripts starting with Gradle 5.1 and a module which has the Java Gradle Plugin applied `apply plugin: 'java-gradle-plugin'` the `signArchives` task fails:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':signArchives'.
> Duplicate key Signature gradle-code-quality-tools-plugin:jar.asc:asc:

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':signArchives'.
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:38)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:49)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:315)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:305)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:101)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:49)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.IllegalStateException: Duplicate key Signature gradle-code-quality-tools-plugin:jar.asc:asc:
        at org.gradle.plugins.signing.Sign.getSignaturesByKey(Sign.java:249)
        at org.gradle.plugins.signing.Sign_Decorated.getSignaturesByKey(Unknown Source)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue$1$1.create(AbstractNestedRuntimeBeanNode.java:82)
        at org.gradle.util.SingleMessageLogger.whileDisabled(SingleMessageLogger.java:341)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue$1.get(AbstractNestedRuntimeBeanNode.java:79)
        at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:167)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue.getValue(AbstractNestedRuntimeBeanNode.java:145)
        at org.gradle.api.internal.tasks.properties.annotations.NestedBeanAnnotationHandler.visitPropertyValue(NestedBeanAnnotationHandler.java:50)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode.visitProperties(AbstractNestedRuntimeBeanNode.java:61)
        at org.gradle.api.internal.tasks.properties.bean.RootRuntimeBeanNode.visitNode(RootRuntimeBeanNode.java:32)
        at org.gradle.api.internal.tasks.properties.DefaultPropertyWalker.visitProperties(DefaultPropertyWalker.java:41)
        at org.gradle.api.internal.tasks.TaskPropertyUtils.visitProperties(TaskPropertyUtils.java:39)
        at org.gradle.api.internal.tasks.execution.DefaultTaskProperties.resolve(DefaultTaskProperties.java:76)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        ... 22 more
```

Are we somehow misusing the API or is this a regression between Gradle 5.1 and Gradle 5.0?

Every recent version before Gradle 5.1 (including Gradle 5.0 & Gradle 4.x) works fine without any problems. Additionally, plain Android/Java/Kotlin modules also work fine.

As for a reproduction sample you can use https://github.com/vanniktech/gradle-maven-publish-plugin bump the Gradle version to 5.1 and execute the `signArchives` task.

**Comment by big-guy** [comment]
I tried upgrading 5.1 in both https://github.com/vanniktech/gradle-maven-publish-plugin and https://github.com/vanniktech/gradle-code-quality-tools-plugin and running signArchives and neither failed because the task was skipped.

I also tried whittling down your example with just the `java-gradle-plugin` and the `signing` plugins applied and it doesn't fail.  Do you have a smaller example that fails?

**Comment by vanniktech** [comment]
Weird. Maybe it's a machine specific problem? Just checking out https://github.com/vanniktech/gradle-maven-publish-plugin/tree/5.1-failing and executing `./gradlew signArchives` at the root level yields the failure:

**Comment by big-guy** [comment]
Is there maybe something user specific that needs to be configured on my side?

**Comment by vanniktech** [comment]
Oh yes. Totally forgot. You need to configure signing:

```groovy
signing.keyId=foo
signing.password=bar
signing.secretKeyRingFile=/home/user/.gnupg/secring.gpg
```

Put these either inside `~/.gradle/gradle.properties` or into a project Gradle property file. If I don't have these configured `signArchives` suceeds for me as well.

**Comment by ndw** [comment]
I can also confirm that this a Gradle 5.1 regression; it works if I back down to 5.0.

**Comment by marcphilipp** [comment]
There seem to be two problems:

1. The `com.vanniktech.maven.publish` unconditionally adds the `jar` task to `configurations.archives.artifacts` where it has already been added by the `java` plugin thus causing the duplicate.
1. The `com.gradle.plugin-publish` plugin adds additional tasks for sources, javadoc, and groovydoc with the same classifiers.

The latter issue can be solved by applying the `com.gradle.plugin-publish` plugin last (because it will then detect that `archives` has already been configured and will not generate additional tasks/archives). For the former I've submitted https://github.com/vanniktech/gradle-maven-publish-plugin/pull/39.

**Comment by koral--** [comment]
> 2. The `com.gradle.plugin-publish` plugin adds additional tasks for sources, javadoc, and groovydoc with the same classifiers.
The latter issue can be solved by applying the com.gradle.plugin-publish plugin last (because it will then detect that archives has already been configured and will not generate additional tasks/archives). 

So, regardless of https://github.com/vanniktech/gradle-maven-publish-plugin/pull/39 if `com.gradle.plugin-publish` is applied before `com.vanniktech.maven.publish` then exception will still be thrown?
If so, maybe `com.vanniktech.maven.publish` should also detect that archives has already been configured and don't do it twice. WDYT?

**Comment by doctorpangloss** [comment]
What's the workaround to this issue? This comes up 100% for all people using gradle and maven publish right now!

**Comment by koral--** [comment]
It should be fixed in gradle-maven-publish-plugin v 0.7.0 https://github.com/vanniktech/gradle-maven-publish-plugin/blob/master/CHANGELOG.md#version-070-2018-01-15

**Comment by vanniktech** [comment]
You do however need to reorder the plugins. My plugin comes first and then the Gradle publish plugin.

**Comment by deanhiller** [comment]
What about the gradle 'signing' plugin.  that hit the same issue on 5.1.1.  I don't have the gradle-maven-publish-plugin at all.  I use the signing plugin and then a sonatype one that depends on the signing one.  This plugin : https://docs.gradle.org/current/userguide/signing_plugin.html

and then after signing is complete, a sonatype plugin..https://github.com/Codearte/gradle-nexus-staging-plugin

Does this https://github.com/vanniktech/gradle-maven-publish-plugin work with sonatype?  I remember there may have been some issue so I changed things to the Codearte one to get my stuff published.

**Comment by manosbatsis** [comment]
For the desperate fellow looking to fix their build (mine was on Gradle 5.4.1), i got away with removing the jar task from `configurations.archives.artifacts`, i.e. changing something like 

```
artifacts { archives jar, sourcesJar, javadocJar }
```

to 

```
artifacts { archives sourcesJar, javadocJar }
```

courtesy of https://github.com/stripe/stripe-java/pull/734
